walkthroughs/week-5-tdd/intro/src/main/java/com/google/sps/Greeter.java

```diff
@@ -22,6 +22,20 @@
    * Returns a greeting for the given name.
    */
   public String greet(String name) {
+    name = name.trim();
+    name = removeSpecialChars(name);
     return "Hello " + name;
   }
+
+  private String removeSpecialChars(String string) {
+    String specialChars = "@#$%";
+    String edittedString = "";
+
+    for (int i = 0; i < string.length(); i++) {
+      if (!specialChars.contains(String.valueOf(string.charAt(i)))) {
+        edittedString += String.valueOf(string.charAt(i));
+      }
+    }
```

Comment:
You should be able to build a regex and use replaceAll to do this for you:

https://docs.oracle.com/javase/7/docs/api/java/lang/String.html#replaceAll(java.lang.String,%20java.lang.String)

Reply:
I actually have a question about this:
I think replaceAll() would be less efficient in this case. Because I would have to call replaceAll() 4 times, on "@", "#", "$", and "%", thus taking more time to compute--since (I think) replaceAll() is syntactic sugar that loops through the input string each time you call it. However, I can replace all 4 special chars by going through the string only once if I do it manually, but the tradeoff would be code complexity (in that the code becomes harder to read because I didn't use a pre-existing function).

So in this case, my question is should we use syntactic sugar for code maintainability/readability, or if we can do the task more efficiently manually, then should we choose that instead?

Reply:
Update: During 1:1 with Brett, he helped me with understanding what I regular expression is, so I think I misunderstood what it's capable of. I'll try using this in a bit and let you know how it goes!

Reply:
Great.

This might not be the best reference but it has what you need: https://www.vogella.com/tutorials/JavaRegularExpressions/article.html

If you can't figure it out then we can give some tips.

Reply:
Update: It worked!